### PR TITLE
Use standardized detection idiom

### PR DESCRIPTION
**Pull request by masterleinad**
Follow-up to https://github.com/dealii/dealii/pull/13285#issuecomment-1020425300, mostly based on https://github.com/kokkos/kokkos/blob/develop/core/src/Kokkos_DetectionIdiom.hpp implementing https://en.cppreference.com/w/cpp/experimental/is_detected.

### CODE REVIEW

File: include/deal.II/base/template_constraints.h [hunk 1]
```diff
@@ -27,6 +27,68 @@
 
 DEAL_II_NAMESPACE_OPEN
 
+template <class...>
+using void_t = void;
```
Comment (bangerth) [inline include/deal.II/base/template_constraints.h:31]: Are we going to use this in public interfaces? If not, can we put it into the internal namespace as well?

Comment (bangerth) [reply include/deal.II/base/template_constraints.h:31]: Actually, all of the things you're adding.

**Review by bangerth — COMMENTED** [review]
This is a neat machine. I like that it's so generically usable. What do others think?

I recognize that this is a minimal patch and as such it's a step in the right direction. I wouldn't mind an eventual follow-up that changes things so that `has_local_element` is not actually a type (namely, either `true_type` or `false_type`, but in fact a boolean. That would not be very difficult to do if we introduced something like
```
  template <...abc...>
  constexpr bool is_detected_v = is_detected<...abc...>::value;
```
I just think that a name like `has_local_element` should actually be a boolean value, not a type :-) But that all can wait for a follow-up.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**